src/main/java/com/yahoo/memory/AllocateDirect.java

```diff
@@ -51,22 +53,25 @@ public void close() {
   private static final class Deallocator implements Runnable {
     //This is the only place the actual native offset is kept for use by unsafe.freeMemory();
     //It can never be modified until it is deallocated.
-    private long actualNativeBaseOffset; //
-    private final ResourceState parentStateRef;
+    private long actualNativeBaseOffset; //set to 0 when deallocated
+    private final long capacity;
+    private final ResourceState myState;
 
     private Deallocator(final ResourceState state) {
       actualNativeBaseOffset = state.getNativeBaseOffset();
+      capacity = state.getCapacity();
       assert (actualNativeBaseOffset != 0);
-      parentStateRef = state;
+      myState = state;
     }
 
     @Override
     public void run() {
       if (actualNativeBaseOffset > 0) {
         unsafe.freeMemory(actualNativeBaseOffset);
+        NioBits.unreserveMemory(capacity);
       }
       actualNativeBaseOffset = 0L;
-      parentStateRef.setInvalid();
+      myState.setInvalid();
```

Comment:
Let's make it the first statement in this method, just in case

Reply:
I think you mean the same as above: unreserve before unsafe.  If so, done.

Reply:
Not exactly. Calling reserveMemory() before actual allocation and unreserveMemory() after actual freeing is required for correctness, when there is not enough memory in the system. reserveMemory()/unreserveMemory() really don't even don't make a lot of sense otherwise.

Here, it's more a matter of style: marking something as invalid before actually making it invalid. From Java Memory Model standpoint, there is no difference, though.

Reply:
Your comment was about the setInvalid()! I wasn't sure what "it" referred to :)